### PR TITLE
[InstCombine] Handle more scalable geps in EmitGEPOffset

### PR DESCRIPTION
**Pull request by davemgreen**
Following up on #71565, this makes scalable splats in EmitGEPOffset use the ElementCount as opposed to assuming it is fixed width, and attempts to handle scalable offsets with vector geps by splatting the vscale to each vector lane. This meant performing the mul by scale separately from CreateVScale.

### CODE REVIEW

File: llvm/lib/Analysis/Local.cpp [hunk 1]
```diff
@@ -66,16 +66,26 @@ Value *llvm::emitGEPOffset(IRBuilderBase *Builder, const DataLayout &DL,
...
-      Value *Scale =
-          !TSize.isScalable() ? ScaleC : Builder->CreateVScale(ScaleC);
+      Value *Scale = ConstantInt::get(IntIdxTy, Size);
+      if (TSize.isScalable()) {
+        Value *VScale;
+        if (IntIdxTy->isVectorTy()) {
+          VScale = Builder->CreateVScale(ConstantInt::get(
+              cast<VectorType>(IntIdxTy)->getElementType(), 1));
+          VScale = Builder->CreateVectorSplat(
+              cast<VectorType>(IntIdxTy)->getElementCount(), VScale);
```
Comment (nikic) [inline llvm/lib/Analysis/Local.cpp:83]: Can we instead move the above splat creation until after the multiply? I.e. calculate on scalars and then splat?

**Comment by davemgreen** [comment]
It appears that the `& PtrSizeMask` can be removed without altering any of the lit tests or any of the programs I tried across AArch64/Arm. Let me know if you think I should re-add it.

**Comment by nikic** [comment]
> It appears that the `& PtrSizeMask` can be removed without altering any of the lit tests or any of the programs I tried across AArch64/Arm. Let me know if you think I should re-add it.

I think this is okay. I confirmed that the APInt uint64_t ctor will implicitly truncate the value, rather than asserting. As such, an explicit mask should not be necessary.

File: llvm/lib/Analysis/Local.cpp [hunk 2]
```diff
@@ -36,25 +36,18 @@ Value *llvm::emitGEPOffset(IRBuilderBase *Builder, const DataLayout &DL,
...
-    uint64_t Size = TSize.getKnownMinValue() & PtrSizeMask;
     if (Constant *OpC = dyn_cast<Constant>(Op)) {
       if (OpC->isZeroValue())
         continue;
 
       // Handle a struct index, which adds its field offset to the pointer.
       if (StructType *STy = GTI.getStructTypeOrNull()) {
         uint64_t OpValue = OpC->getUniqueInteger().getZExtValue();
-        Size = DL.getStructLayout(STy)->getElementOffset(OpValue);
+        unsigned Size = DL.getStructLayout(STy)->getElementOffset(OpValue);
```
Comment (nikic) [inline llvm/lib/Analysis/Local.cpp:50]:
```suggestion
        uint64_t Size = DL.getStructLayout(STy)->getElementOffset(OpValue);
```

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**